File: app/src/main/java/com/ltst/instagramgallerysample/GalleryAdapter.java
package com.ltst.instagramgallerysample;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.cosic.instagallery.data.GalleryData;

import java.util.ArrayList;
import java.util.List;

public class GalleryAdapter extends RecyclerView.Adapter<GalleryAdapter.GalleryViewHolder> {

    private LayoutInflater mInflater;
    private Context mContext;
    private List<GalleryData> mItems = new ArrayList<>();
    private List<Integer> mColors = new ArrayList<>();
    @Nullable
    private OnGalleryClickListener mOnGalleryClickListener;

    public GalleryAdapter(final Context context) {
        mContext = context;
        mInflater = LayoutInflater.from(context);
        for (final String s : mContext.getResources().getStringArray(R.array.default_color_choice_values)) {
            mColors.add(Color.parseColor(s));
        }
    }

    public void setOnGalleryClickListener(@Nullable final OnGalleryClickListener listener) {
        mOnGalleryClickListener = listener;
    }

    public void clear() {
        mItems.clear();
    }

    public void addAll(final List<GalleryData> items) {
        mItems.addAll(items);
    }

    public GalleryData getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public GalleryViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        View view = mInflater.inflate(R.layout.gallery_item_view, parent, false);
        final GalleryViewHolder viewHolder = new GalleryViewHolder(view);
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                if (mOnGalleryClickListener != null) {
                    int adapterPosition = viewHolder.getAdapterPosition();
                    GalleryData item = getItem(adapterPosition);
                    mOnGalleryClickListener.onClick(adapterPosition, item);
                }
            }
        });
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(final GalleryViewHolder holder, final int position) {
        GalleryData item = getItem(position);
        View view = holder.getView();
        View viewById = view.findViewById(R.id.item_view);
        TextView textView = view.findViewById(R.id.item_value);
        viewById.setBackgroundColor(getColor(position));
        textView.setText("" + item.value);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    @Override
    public long getItemId(final int position) {
        return getItem(position).hashCode();
    }

    private int getColor(int position) {
        int i = position % mColors.size();
        return mColors.get(i);
    }

    public static final class GalleryViewHolder extends RecyclerView.ViewHolder {

        public GalleryViewHolder(final View itemView) {
            super(itemView);
        }

        public View getView() {
            return itemView;
        }
    }

    public interface OnGalleryClickListener {
        void onClick(int position, GalleryData item);
    }
}

File: app/src/main/java/com/ltst/instagramgallerysample/GalleryAdapter.kt
package com.ltst.instagramgallerysample

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import java.util.ArrayList

class GalleryAdapter(context: Context) : RecyclerView.Adapter<GalleryAdapter.GalleryViewHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val items = ArrayList<GalleryData>()
    private val colors = ArrayList<Int>()
    private var onGalleryClickListener: OnGalleryClickListener? = null

    init {
        for (s in context.resources.getStringArray(R.array.default_color_choice_values)) {
            colors.add(Color.parseColor(s))
        }
    }

    fun setOnGalleryClickListener(listener: OnGalleryClickListener?) {
        onGalleryClickListener = listener
    }

    fun clear() {
        items.clear()
    }

    fun addAll(items: List<GalleryData>) {
        this.items.addAll(items)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GalleryViewHolder {
        val view = inflater.inflate(R.layout.gallery_item_view, parent, false)
        val viewHolder = GalleryViewHolder(view)
        view.setOnClickListener {
            if (onGalleryClickListener != null) {
                val adapterPosition = viewHolder.adapterPosition
                val item = getItem(adapterPosition)
                onGalleryClickListener!!.onClick(adapterPosition, item)
            }
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: GalleryViewHolder, position: Int) {
        val item = getItem(position)
        val view = holder.view
        val viewById = view.findViewById<View>(R.id.item_view)
        val textView = view.findViewById<TextView>(R.id.item_value)
        viewById.setBackgroundColor(getColor(position))
        textView.text = item.value.toString()
    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).hashCode().toLong()
    }

    private fun getItem(position: Int): GalleryData {
        return items[position]
    }

    private fun getColor(position: Int): Int {
        val i = position % colors.size
        return colors[i]
    }

    class GalleryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val view: View
            get() = itemView
    }

    interface OnGalleryClickListener {
        fun onClick(position: Int, item: GalleryData)
    }
}
